File: figures/src/r/classes/pathway_enrichment_dot.R
repo_root <- getOption("manuscript.figure.repo_root")
if (is.null(repo_root)) {
  stop("manuscript.figure.repo_root option is required")
}
source(file.path(repo_root, "figures", "src", "r", "common.R"))

create_plot <- function(spec_path) {
  spec <- load_yaml(spec_path)
  theme <- load_yaml(file.path(repo_root, "figures/config/project_theme.yml"))
  font_policy <- load_yaml(file.path(repo_root, "figures/config/font_policy.yml"))
  profiles <- load_yaml(file.path(repo_root, "figures/config/venue_profiles.yml"))
  registry <- load_yaml(file.path(repo_root, "figures/registry/classes.yml"))$classes
  profile <- profiles$profiles[[spec$target_profile]]
  class_entry <- registry[[spec$class_id]]
  validate_common_contract(spec, theme, font_policy, profile, class_entry)
  resolved_font <- resolve_font(font_policy)
  plot_theme <- build_theme(theme, resolved_font)

  rows <- read.csv(resolve_data_input(spec, 1), stringsAsFactors = FALSE, check.names = FALSE)
  rows <- rows[order(rows$gene_ratio, decreasing = TRUE), ]
  if (nrow(rows) > 0) {
    rows$pathway <- factor(rows$pathway, levels = rev(rows$pathway))
  } else {
    rows <- data.frame(
      pathway = character(),
      gene_ratio = numeric(),
      neg_log10_fdr = numeric(),
      gene_count = integer(),
      direction = character(),
      highlight_order = integer(),
      stringsAsFactors = FALSE
    )
  }
  write_source_csv(
    file.path(repo_root, source_data_mapping(spec)$a),
    rows,
    c("pathway", "gene_ratio", "neg_log10_fdr", "gene_count", "direction", "highlight_order")
  )

  if (nrow(rows) > 0) {
    plot <- ggplot2::ggplot(rows, ggplot2::aes(x = gene_ratio, y = pathway, color = direction, size = gene_count)) +
      ggplot2::geom_point(alpha = 0.85) +
      ggplot2::scale_color_manual(values = c("up" = theme$palette$categorical[[2]], "down" = theme$palette$categorical[[1]])) +
      ggplot2::geom_text(
        ggplot2::aes(label = sprintf("%.1f", neg_log10_fdr), x = gene_ratio + 0.01),
        hjust = 0,
        size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
        color = theme$palette$neutral[[2]]
      ) +
      ggplot2::labs(title = "Pathway enrichment dot plot", x = "Gene ratio", y = NULL) +
      plot_theme +
      ggplot2::theme(
        plot.tag = ggplot2::element_text(size = as.numeric(theme$panel_labels$font_size_pt), face = theme$panel_labels$font_weight, family = resolved_font$family)
      )
  } else {
    plot <- ggplot2::ggplot() +
      ggplot2::annotate(
        "text",
        x = 0.5,
        y = 0.5,
        label = "No enriched pathways\nfor current fgsea profile",
        size = as.numeric(theme$typography$annotation_font_size_pt) / 2.845,
        color = theme$palette$neutral[[2]],
        family = resolved_font$family
      ) +
      ggplot2::coord_cartesian(xlim = c(0, 1), ylim = c(0, 1), clip = "off") +
      ggplot2::labs(title = "Pathway enrichment dot plot", x = "Gene ratio", y = NULL) +
      plot_theme +
      ggplot2::theme(
        axis.text.y = ggplot2::element_blank(),
        axis.ticks.y = ggplot2::element_blank(),
        plot.tag = ggplot2::element_text(size = as.numeric(theme$panel_labels$font_size_pt), face = theme$panel_labels$font_weight, family = resolved_font$family)
      )
  }
  plot <- plot + patchwork::plot_annotation(tag_levels = "a")
  list(
    plot = plot,
    spec = spec,
    theme = theme,
    profile = profile,
    resolved_font = resolved_font,
    rows = rows
  )
}

build_figure <- function(spec_path) {
  payload <- create_plot(spec_path)
  spec <- payload$spec
  theme <- payload$theme
  profile <- payload$profile
  resolved_font <- payload$resolved_font
  plot <- payload$plot
  rows <- payload$rows
  output_dir <- file.path(repo_root, spec$renderers$r$output_dir)
  dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
  stem <- spec$figure_id
  svg_path <- file.path(output_dir, sprintf("%s.svg", stem))
  pdf_path <- file.path(output_dir, sprintf("%s.pdf", stem))
  png_path <- file.path(output_dir, sprintf("%s.png", stem))
  width_in <- mm_to_inches(as.numeric(spec$size$width_mm))
  height_in <- mm_to_inches(as.numeric(spec$size$height_mm))

  svglite::svglite(file = svg_path, width = width_in, height = height_in, bg = theme$export_defaults$background, system_fonts = list(sans = resolved_font$family))
  print(plot)
  grDevices::dev.off()
  grDevices::cairo_pdf(file = pdf_path, width = width_in, height = height_in, bg = theme$export_defaults$background, family = resolved_font$family)
  print(plot)
  grDevices::dev.off()
  ragg::agg_png(filename = png_path, width = width_in, height = height_in, units = "in", res = as.numeric(profile$preview_dpi), background = theme$export_defaults$background)
  print(plot)
  grDevices::dev.off()

  outputs <- list(svg = sub(paste0("^", repo_root, "/"), "", svg_path), pdf = sub(paste0("^", repo_root, "/"), "", pdf_path), png = sub(paste0("^", repo_root, "/"), "", png_path))
  write_manifest(
    repo_root,
    spec,
    profile,
    resolved_font,
    spec_path,
    "r",
    outputs,
    list(
      "dot_size_encodes_gene_count",
      "annotation_for_significance_strength",
      "direction_color_encoding",
      "vector_first_export"
    ),
    list(
      panel_count = length(spec$panels),
      annotation_count = nrow(rows)
    )
  )
  outputs
}

File: figures/src/python/classes/pathway_enrichment_dot.py
#!/usr/bin/env python3
"""Class-based Python renderer for pathway-enrichment dot plots."""

from __future__ import annotations

import csv
from pathlib import Path
from typing import Any

from matplotlib.figure import Figure

from figures.src.python.common import (
    REPO_ROOT,
    apply_publication_layout,
    configure_matplotlib,
    export_csv,
    mm_to_inches,
    project_resources,
    resolve_data_input,
    resolve_font_stack,
    source_data_mapping,
    validate_common_contract,
    write_manifest,
)


def _load_rows(path: Path) -> list[dict[str, Any]]:
    with path.open("r", encoding="utf-8", newline="") as handle:
        reader = csv.DictReader(handle)
        return [
            {
                "pathway": row["pathway"],
                "gene_ratio": float(row["gene_ratio"]),
                "neg_log10_fdr": float(row["neg_log10_fdr"]),
                "gene_count": int(row["gene_count"]),
                "direction": row["direction"],
                "highlight_order": int(row["highlight_order"]),
            }
            for row in reader
        ]


def build_source_data(spec: dict[str, Any], rows: list[dict[str, Any]]) -> list[dict[str, Any]]:
    ordered = sorted(rows, key=lambda item: item["gene_ratio"], reverse=True)
    export_csv(
        REPO_ROOT / source_data_mapping(spec)["a"],
        ordered,
        ["pathway", "gene_ratio", "neg_log10_fdr", "gene_count", "direction", "highlight_order"],
    )
    return ordered


def create_figure(spec_path: Path) -> Figure:
    spec, theme, font_policy, profile, class_entry = project_resources(spec_path)
    validate_common_contract(spec, theme, font_policy, profile, class_entry)
    rows = build_source_data(spec, _load_rows(resolve_data_input(spec, 0)))
    resolved_font = resolve_font_stack(font_policy)
    configure_matplotlib(theme, font_policy, resolved_font)

    import matplotlib.pyplot as plt

    fig, axis = plt.subplots(
        1,
        1,
        figsize=(
            mm_to_inches(float(spec["size"]["width_mm"])),
            mm_to_inches(float(spec["size"]["height_mm"])),
        ),
        constrained_layout=bool(theme["layout"]["constrained"]),
    )
    apply_publication_layout(fig, theme)
    colors = {"up": theme["palette"]["categorical"][1], "down": theme["palette"]["categorical"][0]}
    y_positions = list(range(len(rows)))
    if rows:
        axis.scatter(
            [row["gene_ratio"] for row in rows],
            y_positions,
            s=[row["gene_count"] * 8 for row in rows],
            c=[colors[row["direction"]] for row in rows],
            alpha=0.85,
            edgecolors=theme["palette"]["neutral"][0],
            linewidths=0.5,
        )
    for idx, row in enumerate(rows):
        axis.text(
            row["gene_ratio"] + 0.01,
            idx,
            f"{row['neg_log10_fdr']:.1f}",
            va="center",
            ha="left",
            fontsize=theme["typography"]["annotation_font_size_pt"],
            color=theme["palette"]["neutral"][1],
        )
    axis.text(
        float(theme["panel_labels"]["x_position_axes"]),
        float(theme["panel_labels"]["y_position_axes"]),
        "a",
        transform=axis.transAxes,
        fontsize=theme["panel_labels"]["font_size_pt"],
        fontweight=theme["panel_labels"]["font_weight"],
        va="top",
    )
    axis.set_yticks(y_positions)
    axis.set_yticklabels([row["pathway"] for row in rows])
    if rows:
        axis.invert_yaxis()
        x_min = min(float(row["gene_ratio"]) for row in rows)
        x_max = max(float(row["gene_ratio"]) for row in rows)
        axis.set_xlim(max(0.0, x_min - 0.03), x_max + 0.12)
        axis.set_ylim(len(rows) - 0.5, -0.9)
    else:
        axis.set_xlim(0.0, 1.0)
        axis.set_ylim(0.0, 1.0)
        axis.set_yticks([])
        axis.set_xticks([0.0, 0.5, 1.0])
        axis.text(
            0.5,
            0.5,
            "No enriched pathways\nfor current fgsea profile",
            transform=axis.transAxes,
            ha="center",
            va="center",
            fontsize=theme["typography"]["annotation_font_size_pt"],
            color=theme["palette"]["neutral"][1],
        )
    axis.set_title("Pathway enrichment dot plot", loc="left", pad=2)
    axis.set_xlabel("Gene ratio")
    axis.set_ylabel("")
    axis.spines["top"].set_visible(False)
    axis.spines["right"].set_visible(False)
    return fig


def build_figure(spec_path: Path) -> dict[str, str]:
    spec, theme, font_policy, profile, _ = project_resources(spec_path)
    rows = _load_rows(resolve_data_input(spec, 0))
    resolved_font = resolve_font_stack(font_policy)
    fig = create_figure(spec_path)
    output_dir = REPO_ROOT / str(spec["renderers"]["python"]["output_dir"])
    output_dir.mkdir(parents=True, exist_ok=True)
    stem = str(spec["figure_id"])
    svg_path = output_dir / f"{stem}.svg"
    pdf_path = output_dir / f"{stem}.pdf"
    png_path = output_dir / f"{stem}.png"
    fig.savefig(svg_path, format="svg")
    fig.savefig(pdf_path, format="pdf")
    fig.savefig(png_path, format="png", dpi=int(profile["preview_dpi"]))
    outputs = {
        "svg": str(svg_path.relative_to(REPO_ROOT)),
        "pdf": str(pdf_path.relative_to(REPO_ROOT)),
        "png": str(png_path.relative_to(REPO_ROOT)),
    }
    write_manifest(
        spec,
        profile,
        resolved_font,
        "python",
        spec_path,
        outputs,
        [
            "dot_size_encodes_gene_count",
            "annotation_for_significance_strength",
            "direction_color_encoding",
            "vector_first_export",
        ],
        {
            "panel_count": len(spec.get("panels", [])),
            "annotation_count": len(rows),
        },
    )
    return outputs


if __name__ == "__main__":
    raise SystemExit("Use figures/src/python/run_class_renderer.py")
